dockstore-client/src/main/java/io/dockstore/client/cli/nested/BaseLanguageClient.java

```diff
@@ -199,6 +212,76 @@ public long launchPipeline(String entryVal, boolean localEntry, String yamlFile,
         return 0;
     }
 
+    /**
+     * Validates the locally downloaded descriptor file has the same SHA-1 checksum as the descriptor stored in the database
+     * @param type CWL or WDL or NFL
+     * @param entryVal Tool/workflow path
+     * @return void errors out if checksums do not match, provides a warning if a remote checksum is null
+     */
+    public void validateDescriptorChecksum(ToolDescriptor.TypeEnum type, String entryVal) {
+        final String[] parts = entryVal.split(":");
+        String path = parts[0];
+
+        // workflows/checkers have a special prefix for TRS endpoints
+        String ga4ghv20Path = path;
+        if (abstractEntryClient.getEntryType().toLowerCase().equals("workflow") || abstractEntryClient.getEntryType().toLowerCase().equals("checker")) {
+            ga4ghv20Path = "#workflow/" + ga4ghv20Path;
+        }
+
+        // Get the entry version we are validating for
+        String versionID = abstractEntryClient.getVersionID(entryVal);
+
+        List<ToolFile> allDescriptors = abstractEntryClient.getAllToolDescriptors(type.toString(), path, versionID);
+
+        Ga4Ghv20Api ga4ghv20api = abstractEntryClient.getClient().getGa4Ghv20Api();
+
+        // All secondary files are located relative to the location of the primary descriptor.
+        final String localTemporaryDirectory = localPrimaryDescriptorFile.getParent() + "/";
+
+        // Validate each tool file associated with the entry (Primary and secondary descriptors)
+        for (ToolFile toolFile : allDescriptors) {
+            Checksum remoteDescriptorChecksum = null;
+            Checksum localDescriptorChecksum = null;
+
+            // Get remote descriptor checksum
+            try {
+                // The TRS endpoint only discovers published entries
+                final FileWrapper remoteDescriptor = ga4ghv20api.toolsIdVersionsVersionIdTypeDescriptorRelativePathGet(type.toString(), ga4ghv20Path, versionID, toolFile.getPath());
+                List<Checksum> remoteChecksumList = remoteDescriptor.getChecksum()
```

Comment:
I kept as-is for now. I think keeping both checksums as the same type is nice. Happy to modify anyways, if the other approach is more standard/preferred, or potentially better?

Reply:
> I kept as-is for now. I think keeping both checksums as the same type is nice. Happy to modify anyways, if the other approach is more standard/preferred, or potentially better?

My opinion on nulls:

They're sometimes unavoidable, e.g., when using 3rd party libraries. And in small blocks of code it may not matter that much. But in general nulls are evil and I prefer to avoid them at all costs. :) If you have a possible null,  you have to check that it's not null. More crucially, you have to remember to check for it. Failures to check for null are probably the most common type of runtime error that occur in Java.

Whereas with Optional, you know you have to check for it for being empty -- the compiler enforces it.

It's more crucial when you return a value from a method that could be null. If you're using it within a method, like here, it's less of an issue.

Reply:
I can always support avoiding NPEs!

Refactored [here](https://github.com/dockstore/dockstore-cli/blob/8e1a149956ea321769c5bc7340e41e081b5bee37/dockstore-client/src/main/java/io/dockstore/client/cli/nested/BaseLanguageClient.java#L243) to use Optional\<Checksum\> for the remote descriptor. Also switched around the checksum comparison block to avoid additional conditional checks.